### PR TITLE
[enhancement](Nereids)refactor sort plan in nereids

### PR DESCRIPTION
**Pull request by morrySnow**
# Proposed changes

- rename PhysicalHeapSort to PhysicalQuickSort
- add LogicalTopN and PhysicalTopN
- add implementation rule for LogicalTopN
- add a interface Sort for both logical and physical sort
- add a interface TopN for both logical and physical top-n
- add a AbstractPhysicalSort as super class of PhysicalQuickSort and PhysicalTopN

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/algebra/TopN.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.trees.plans.algebra;
+
+/**
+ * Common interface for logical/physical TopN.
+ */
+public interface TopN extends Sort {
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/algebra/TopN.java:23]: We already have limit and sort. Why we need TopN? What is the difference between TopN and a combination of sort and limit?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalQuickSort.java [hunk 2]
```diff
@@ -30,56 +30,30 @@
...
 import java.util.List;
-import java.util.Objects;
 import java.util.Optional;
 
 /**
- * Physical sort plan.
+ * Physical quick sort plan.
  */
-public class PhysicalHeapSort<CHILD_TYPE extends Plan> extends PhysicalUnary<CHILD_TYPE> {
+public class PhysicalQuickSort<CHILD_TYPE extends Plan> extends AbstractPhysicalSort<CHILD_TYPE> {
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalQuickSort.java:38]: Does "QuickSort" here stand for quick sort algorithm? It seems that this node is irrelevant to sort algorithm chosen by BE.

**Comment by morrySnow** [comment]
@englefly @jackwener PTAL

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalQuickSort.java (hunk 2 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalQuickSort.java:38]: without limit and offset, be use quick sort to do sort indeed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/algebra/TopN.java (hunk 1 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/algebra/TopN.java:23]: yes, we will translate limit(sort) to topN when exploration. and look up witch has best cost

**Review by englefly — DISMISSED** [review]

**Comment by morrySnow** [comment]
@924060929 PTAL

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 3]
```diff
@@ -286,10 +288,42 @@ public PlanFragment visitLogicalSort(LogicalSort<Plan> sort, PlanTranslatorConte
     }
 
     @Override
-    public PlanFragment visitPhysicalHeapSort(PhysicalHeapSort<Plan> sort,
+    public PlanFragment visitPhysicalQuickSort(PhysicalQuickSort<Plan> sort,
             PlanTranslatorContext context) {
+        PlanFragment childFragment = visitAbstractPhysicalSort(sort, context);
+        SortNode sortNode = (SortNode) childFragment.getPlanRoot();
+        //isPartitioned()==true means there is only one instance, so no merge phase
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:295]: isPartitioned() == false

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 4]
```diff
@@ -286,10 +288,42 @@ public PlanFragment visitLogicalSort(LogicalSort<Plan> sort, PlanTranslatorConte
...
+        return mergeFragment;
+    }
+
+    @Override
+    public PlanFragment visitPhysicalTopN(PhysicalTopN<Plan> topN, PlanTranslatorContext context) {
+        PlanFragment childFragment = visitAbstractPhysicalSort(topN, context);
+        SortNode sortNode = (SortNode) childFragment.getPlanRoot();
+        sortNode.setOffset(topN.getOffset());
+        sortNode.setLimit(topN.getLimit());
+        //isPartitioned()==true means there is only one instance, so no merge phase
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:312]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitor.java [hunk 5]
```diff
@@ -130,10 +137,18 @@ public R visitPhysicalOlapScan(PhysicalOlapScan olapScan, C context) {
         return visitPhysicalScan(olapScan, context);
     }
 
-    public R visitPhysicalHeapSort(PhysicalHeapSort<Plan> sort, C context) {
+    public R visitAbstractPhysicalSort(AbstractPhysicalSort<Plan> sort, C context) {
         return visit(sort, context);
     }
 
+    public R visitPhysicalQuickSort(PhysicalQuickSort<Plan> sort, C context) {
+        return visit(sort, context);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitor.java:145]:
```suggestion
        return visitAbstractPhysicalSort(sort, context);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/AbstractPhysicalSort.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * Constructor of PhysicalHashJoinNode.
+     */
+    public AbstractPhysicalSort(PlanType type, List<OrderKey> orderKeys,
+            Optional<GroupExpression> groupExpression, LogicalProperties logicalProperties,
+            CHILD_TYPE child) {
+        super(type, groupExpression, logicalProperties, child);
+        this.orderKeys = orderKeys;
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/AbstractPhysicalSort.java:53]: Objects.requireNonNull(orderKeys)

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalTopN.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        return offset;
+    }
+
+    public int getLimit() {
+        return limit;
+    }
+
+    @Override
+    public String toString() {
+        return "LogicalTopN (" + StringUtils.join(orderKeys, ", ") + ")";
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalTopN.java:81]: Not add offset/limit ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalTopN.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        LogicalTopN that = (LogicalTopN) o;
+        return this.offset == that.offset && this.limit == that.limit && Objects.equals(this.orderKeys, that.orderKeys);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(orderKeys, offset);
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalTopN.java:98]: limit?

**Review by 924060929 — APPROVED** [review]
LGTM

**Pull request merged**